src/core/Akka.Cluster.Tests/MemberOrderingModelBasedTests.cs

```diff
@@ -0,0 +1,250 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="MemberOrderingModelBasedTests.cs" company="Akka.NET Project">
+//     Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+//     Copyright (C) 2013-2016 Akka.NET project <https://github.com/akkadotnet/akka.net>
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Akka.Actor;
+using Akka.Tests.Shared.Internals.Helpers;
+using Akka.Util.Internal;
+using FsCheck;
+using FsCheck.Experimental;
+using FsCheck.Xunit;
+
+namespace Akka.Cluster.Tests
+{
+    public class MemberOrderingModelBasedTests
+    {
+        public MemberOrderingModelBasedTests()
+        {
+            // register the custom generators to make testing easier
+            Arb.Register<ClusterGenerators>();
+        }
+
+        [Property]
+        public Property MemberOrderingMustObeyModel()
+        {
+            return new MembershipMachine().ToProperty();
+        }
+
+        [Property(MaxTest = 1000)]
+        public Property DistinctMemberAddressesMustCompareDifferently(Address[] addresses)
+        {
+            if (addresses.Length == 0)
+                return true.Classify(true, "empty set");
+            var addressCount = addresses.Length;
+            var distinctCount = addresses.Distinct().Count(); // uses `Address.Equals` under the hood
+            if (addressCount != distinctCount)
+                return true.Classify(false, "all distinct").Classify(false, "empty set");
+
+            // use the sort to help us short-circuit a comparison problem
+            var sortedAddresses = addresses.OrderBy(x => x, Member.AddressOrdering).ToList();
+
+            var a1 = sortedAddresses.First();
+            foreach (var a2 in sortedAddresses.Skip(1))
+            {
+                if (Member.AddressOrdering.Compare(a1, a2) == 0)
+                    return false.Classify(true, "all distinct").Classify(false, "empty set")
+                        .Label($"Expected {a1} and {a2} to compare as different, but did not. Compared as same");
+                a1 = a2; // next member
+            }
+
+            return true.Classify(true, "all distinct").Classify(false, "empty set"); ;
+        }
+    }
+
+    public class MembershipMachine : Machine<MembershipState, MembershipModel>
```

Comment:
Heavy-duty model-based test for squishing the issues with collections and `Member` comparisons I outlined here: https://github.com/akkadotnet/akka.net/issues/2584#issuecomment-310974051